docs(execution_environments): add best practices for building and managing EEs (#188)
redhat-cop/automation-good-practices: execution_environments/README.adoc

EEs are built using https://ansible.readthedocs.io/projects/builder/en/latest/[`ansible-builder`], which reads an `execution-environment.yml` definition file and produces a container image.
This section covers good practices for defining, building, and maintaining execution environments.

For good practices on declaring EEs as objects in Automation Controller configuration, see <<aap_configuration/README.adoc#_aap_configuration_as_code,AAP Configuration as Code>>.
For good practices on immutable references to EE images in production, see <<git_workflow/README.adoc#_never_use_latest_or_branch_names_in_production_references,Git Workflow: Never use "latest" or branch names in production references>>.
For vulnerability scanning of built EE images, see <<security/README.adoc#_scan_container_images_and_generate_sboms_for_every_release,Security: Scan container images and generate SBOMs for every release>>.

== Always use `version: 3` in the `execution-environment.yml` definition file
[%collapsible]
====
Explanations::

> The `execution-environment.yml` definition file must specify `version: 3` as its first key.

Reviewer: YAML is unordered so there is no explicit requirement that `version: 3` must be the first key in the file

----
# execution-environment.yml
version: 3

images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:2.16

dependencies:
  galaxy:
    collections:
      - name: redhat.rhel_system_roles
        version: "1.23.0"
----

.Don't do this:
[source,yaml]
----
# Missing version: 3 -- ansible-builder defaults to the v1 schema
# Using any v3-only key (images:, options:, additional_build_files:, etc.)
# causes an immediate DefinitionError: "Additional properties are not allowed"

images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:2.16

dependencies:
  galaxy:
    collections:
      - name: redhat.rhel_system_roles
        version: "1.23.0"
----
====

== Use `ee-minimal` as the base image for custom execution environments
[%collapsible]
====
Explanations::
Red Hat provides two categories of supported base images:
+
* `ee-minimal` -- contains only `ansible-core` and `ansible-runner`.
It is the recommended starting point for custom EEs.
* `ee-supported` -- includes a curated set of pre-installed certified collections.
It is designed to be used as-is in environments where its bundled content is sufficient.

> Using it as a base for custom images and then specifying additional or different versions of the same collections creates conflicts that are difficult to diagnose.

Reviewer: ```suggestion
Using it as a base for custom images and then specifying additional or different versions of the same collections often creates conflicts that are difficult to diagnose.
```

+
An EE built on `ee-minimal` is smaller, faster to pull, easier to reason about, and carries no transitive dependencies that were not explicitly chosen.

.Do this:
[source,yaml]
----
version: 3

images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:2.16

dependencies:
  galaxy:
    collections:
      - name: redhat.rhel_system_roles
        version: "1.23.0"
      - name: ansible.posix
        version: "1.5.4"
----

.Don't do this:
[source,yaml]
----
version: 3

# ee-supported already bundles a fixed version of amazon.aws (and many others).
# Adding a different version here creates a conflict: two copies of the same
# collection at different versions may be present in the image, leading to
# unpredictable behavior at runtime.
images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-supported-rhel9:2.16

dependencies:
  galaxy:
    collections:
      - name: amazon.aws
        version: "8.0.0"
----
====

== Always pin the base image to a specific version -- never use `:latest`
[%collapsible]
====
Explanations::

> The base image tag in the `images.base_image.name` field must always reference a specific version, not the `:latest` shorthand.

Reviewer: ```suggestion
The base image tag in the `images.base_image.name` field must always reference a specific tag version, and must not use `:latest` as a value.
```

+
Red Hat distributes EE base images under two naming schemes, both requiring an explicit version:
+
[cols="1,2,2"]
|===
| Type | Example | Use case

| *Versioned* (tied to an AAP release)

> | `registry.redhat.io/ansible-automation-platform-27/ee-minimal-rhel9:2.7.0`

Reviewer: This is not resolvable

----

[source,yaml]
----
version: 3

# Option B -- version-less tag (tracks ansible-core 2.16.x)
images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:2.16
----

[source,yaml]
----
version: 3

# Option C -- SHA digest for maximum reproducibility
# Obtain the digest with: skopeo inspect docker://registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:2.16 | jq -r .Digest
images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9@sha256:<digest>
----

.Don't do this:
[source,yaml]
----
version: 3

# :latest was removed in AAP 2.7 -- this will fail to build
images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:latest
----
====

== Do not create monolithic execution environments -- group by automation domain
[%collapsible]
====
Explanations::
Each EE should contain only the dependencies required by a specific domain of automation.
Common domain groupings include:
+
* Linux / RHEL system management
* Windows management
* Networking (Cisco, Juniper, Arista, etc.)
* Cloud (AWS, Azure, GCP)
* Virtualization (VMware, OpenShift Virt)
+

> A separate, lightweight EE should be created for each domain rather than a single large image containing all collections.

Reviewer: ```suggestion
A separate, lightweight EE should be created for each domain rather than a single large image containing all collections and dependencies.
```

Rationale::
Combining unrelated automation workloads into a single EE creates several problems:
+
* *Dependency conflicts* -- networking collections may require specific Python library versions that conflict with those required by cloud collections.
* *Image bloat* -- unused collections increase image size, slowing pulls and increasing storage costs in the registry.

> * *Maintenance burden* -- a change in any dependency forces a rebuild and validation of the entire image, even for unrelated automation.

Reviewer: ```suggestion
* *Maintenance burden* -- a change in any dependency forces a rebuild and revalidation of the entire image, even for unrelated automation.
```

----

.Don't do this:
[source,yaml]
----
# ee-everything.yml -- single monolithic EE with unrelated domains
version: 3
images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:2.16
dependencies:
  galaxy:
    collections:
      - name: redhat.rhel_system_roles    # Linux
        version: "1.23.0"
      - name: cisco.ios                   # Networking
        version: "10.1.1"
      - name: amazon.aws                  # Cloud
        version: "8.0.0"
      - name: community.windows           # Windows
        version: "2.3.0"
----
====

== Pin all dependency versions explicitly
[%collapsible]
====
Explanations::
Every dependency referenced in an EE definition must be pinned to an exact version:
+
* Collections in `dependencies.galaxy`: use the `version:` key with an exact version string (e.g., `"1.23.0"`).
* Python packages in `dependencies.python`: use `==` pinning (e.g., `jmespath==1.0.1`).
* System packages in `dependencies.system`: use the specific package name; avoid wildcards.
* `ansible-core` and `ansible-runner` in `dependencies.ansible_core` / `dependencies.ansible_runner`: pin with `package_pip: ansible-core==<version>`.
+

> An external `requirements.yml` or `requirements.txt` file is equally valid, but the same pinning requirement applies to those files.

Reviewer: Any guidance between defining dependencies inline instead of in external files

> An external `requirements.yml` or `requirements.txt` file is equally valid, but the same pinning requirement applies to those files.

Author: Good question — we intentionally left this as 'either is valid' without a strong preference, but it may be worth adding explicit guidance. A few trade-offs worth discussing:

**Inline (inside `execution-environment.yml`):**
- Single file to track; easier to review as a unit
- Ideal for simple EEs with a small number of dependencies
- Version pins are immediately visible alongside the image and options config

**External files (`requirements.yml`, `requirements.txt`, `bindep.txt`):**
- Allows reuse across multiple EE definitions that share a common dependency set
- More familiar to users coming from classic Ansible workflows (requirements.yml is already a known pattern)
- Easier to integrate with tooling that expects separate requirements files

A reasonable recommendation might be: **prefer inline for new EEs** (keeps everything in one place, reduces file count), and **use external files when dependencies are shared across multiple EE definitions** or when integrating with existing tooling that expects them separately.

Happy to add a short note to this effect if the maintainers agree this guidance is useful here.

.Do this:
[source,yaml]
----
version: 3

images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:2.16

options:
  package_manager_path: /usr/bin/microdnf   # required for ee-minimal

dependencies:
  galaxy:
    collections:
      - name: redhat.rhel_system_roles
        version: "1.23.0"
  system:
    - git-core [platform:rpm]
    - python3-devel [platform:rpm]
----

.Don't do this:
[source,yaml]
----
version: 3

images:
  base_image:
    name: registry.redhat.io/ansible-automation-platform/ee-minimal-rhel9:2.16

# Missing options.package_manager_path -- build will fail on system deps
# because microdnf is not at /usr/bin/dnf

dependencies:
  galaxy:
    collections:
      - name: redhat.rhel_system_roles
        version: "1.23.0"
  system:
    - git-core [platform:rpm]
----
====

== Store EE definitions in Git and automate builds through a CI/CD pipeline
[%collapsible]
====
Explanations::
Every `execution-environment.yml` file, along with any supporting `requirements.yml`, `requirements.txt`, and `bindep.txt` files, must be stored in a Git repository.
+

> Builds must be triggered automatically by CI/CD pipelines on relevant Git events (push to a watched branch, pull request merge, or a scheduled trigger for regular security updates).

Reviewer: ```suggestion
Builds must be triggered automatically by CI/CD pipelines on relevant Git events (push to a watched branch, pull requests, or a scheduled trigger for regular security updates).
```
